Quadrotor runDynamics ported to System2.0

foehnph opened the pull request:
Addresses issue #3994 Spiral 1
Simulates the quadrotor without inputs in a warehouse environment.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/3995)
<!-- Reviewable:end -->

naveenoid commented:
Yikes! Jenkins spam! Anyways,
+@liangfok for feature review.
+@sammy-tri for platform review.

jwnimmer-tri commented:
@drake-jenkins-bot ok to test

@naveenoid The branch has conflicts.  Resolve those before seeking review.

naveenoid commented:
Marked as do-not-review until @foehnph can fix the conflicts.

liangfok commented:
I just left a few nits after a quick skim. I'm concerned about the deleted externals. I doubt they should be deleted.

---

Reviewed 1 of 12 files at r1.
Review status: 1 of 12 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[drake/examples/Quadrotor/CMakeLists.txt, line 5 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KV_BnWugJA2bNK7-vnf:-KV_BnWugJA2bNK7-vng:b3cp2yw) ([raw file](https://github.com/robotlocomotion/drake/blob/088cee067bf7e241e6e4298bdd9cb6adcd882aa5/drake/examples/Quadrotor/CMakeLists.txt#L5)):*
> ```CMake
> if(lcm_FOUND AND NOT (WIN32 AND (CMAKE_SIZEOF_VOID_P EQUAL 4)))
>   add_executable(runQuadrotorDynamics runDynamics.cpp)
>   target_link_libraries(runQuadrotorDynamics drakeRigidBodyPlant drakeRBSystem drakeLCMSystem2 gflags )
> ```

I don't think `drakeRBSystem` is needed anymore.

---

*[drake/examples/Quadrotor/runDynamics.cpp, line 3 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KV_BkEtn091WardkjPe:-KV_BkEtn091WardkjPf:bcsdjpm) ([raw file](https://github.com/robotlocomotion/drake/blob/088cee067bf7e241e6e4298bdd9cb6adcd882aa5/drake/examples/Quadrotor/runDynamics.cpp#L3)):*
> ```cpp
> #include <gflags/gflags.h>
> #include "drake/common/text_logging.h"
> #include <drake/systems/plants/RigidBodySystem.h>
> ```

I don't think this is needed anymore.

---

*[drake/examples/Quadrotor/runDynamics.cpp, line 11 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KV_Bxla2OTHcpSWllEr:-KV_Bxla2OTHcpSWllEs:b-4h0rp3) ([raw file](https://github.com/robotlocomotion/drake/blob/088cee067bf7e241e6e4298bdd9cb6adcd882aa5/drake/examples/Quadrotor/runDynamics.cpp#L11)):*
> ```cpp
> #include "drake/systems/framework/primitives/constant_vector_source.h"
> #include "drake/systems/plants/parser_urdf.h"
> #include "drake/systems/plants/parser_sdf.h"
> ```

Alpha sort.

---

*[drake/examples/Quadrotor/runDynamics.cpp, line 64 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KV_CB6dKUkeaZ2_6p0T:-KV_CB6dKUkeaZ2_6p0U:byvyoa2) ([raw file](https://github.com/robotlocomotion/drake/blob/088cee067bf7e241e6e4298bdd9cb6adcd882aa5/drake/examples/Quadrotor/runDynamics.cpp#L64)):*
> ```cpp
>     Eigen::Isometry3d T_element_to_link = Eigen::Isometry3d::Identity();
>     T_element_to_link.translation() << 0, 0,
>         -kBoxDepth / 2.0;  // top of the box is at z = 0
> ```

"top" should be "Top". Missing period at end of sentence.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-:-KV_CMVWyxSQ4yGPrFX7:b1xazb9)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
@liangfok, I think @foehnph managed to fix the externals issue. I am now reopening for review since the rebase looks ok.

foehnph commented:
@liangfok fixed the defects. PTAL

---

Review status: 0 of 2 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/examples/Quadrotor/CMakeLists.txt, line 5 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KV_BnWugJA2bNK7-vnf:-KV_HoAVdvBGANUvh3d7:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/088cee067bf7e241e6e4298bdd9cb6adcd882aa5/drake/examples/Quadrotor/CMakeLists.txt#L5)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I don't think `drakeRBSystem` is needed anymore.
</details>
Done.

---

*[drake/examples/Quadrotor/runDynamics.cpp, line 3 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KV_BkEtn091WardkjPe:-KV_GcsKLXBir5EhPeHi:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/088cee067bf7e241e6e4298bdd9cb6adcd882aa5/drake/examples/Quadrotor/runDynamics.cpp#L3)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I don't think this is needed anymore.
</details>
Done.

---

*[drake/examples/Quadrotor/runDynamics.cpp, line 11 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KV_Bxla2OTHcpSWllEr:-KV_GJ88LyBtFz81mV3J:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/088cee067bf7e241e6e4298bdd9cb6adcd882aa5/drake/examples/Quadrotor/runDynamics.cpp#L11)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Alpha sort.
</details>
Done.

---

*[drake/examples/Quadrotor/runDynamics.cpp, line 64 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KV_CB6dKUkeaZ2_6p0T:-KV_HcVc_ga-p49xhm38:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/088cee067bf7e241e6e4298bdd9cb6adcd882aa5/drake/examples/Quadrotor/runDynamics.cpp#L64)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> "top" should be "Top". Missing period at end of sentence.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-:-KV_IStV8nn5XbqCs8qA:b-ffb81k)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I completed my first round of feature review. Looking good. PTAL.

---

Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/examples/Quadrotor/CMakeLists.txt, line 5 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KV_SIGsURj-Ick5Cjtj:-KV_SIGsURj-Ick5Cjtk:b-wzr2bo) ([raw file](https://github.com/robotlocomotion/drake/blob/483c97975c17db73809d5cd07fd27d78421cb489/drake/examples/Quadrotor/CMakeLists.txt#L5)):*
> ```CMake
> if(lcm_FOUND AND NOT (WIN32 AND (CMAKE_SIZEOF_VOID_P EQUAL 4)))
>   add_executable(runQuadrotorDynamics runDynamics.cpp)
>   target_link_libraries(runQuadrotorDynamics drakeRigidBodyPlant drakeLCMSystem2 gflags )
> ```

BTW: nit: Extra space before closing ")".

---

*[drake/examples/Quadrotor/CMakeLists.txt, line 6 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KV_SMFMiAZcd_YeNZKA:-KV_SMFMiAZcd_YeNZKB:b-5wqmeb) ([raw file](https://github.com/robotlocomotion/drake/blob/483c97975c17db73809d5cd07fd27d78421cb489/drake/examples/Quadrotor/CMakeLists.txt#L6)):*
> ```CMake
>   add_executable(runQuadrotorDynamics runDynamics.cpp)
>   target_link_libraries(runQuadrotorDynamics drakeRigidBodyPlant drakeLCMSystem2 gflags )
>   drake_add_test(NAME runQuadrotorDynamics COMMAND runQuadrotorDynamics --duration=1.0 SIZE mediu)
> ```

Typo: "mediu"

---

*[drake/examples/Quadrotor/runDynamics.cpp, line 1 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KV_SfXedIGWPOVewdHB:-KV_SfXedIGWPOVewdHC:bouq182) ([raw file](https://github.com/robotlocomotion/drake/blob/483c97975c17db73809d5cd07fd27d78421cb489/drake/examples/Quadrotor/runDynamics.cpp#L1)):*
> ```cpp
> #include <iostream>
> ```

BTW: Please confirm that this is required.

---

*[drake/examples/Quadrotor/runDynamics.cpp, line 2 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KV_SlFP1LiqOYdyO7MF:-KV_SlFP1LiqOYdyO7MG:bouq182) ([raw file](https://github.com/robotlocomotion/drake/blob/483c97975c17db73809d5cd07fd27d78421cb489/drake/examples/Quadrotor/runDynamics.cpp#L2)):*
> ```cpp
> #include <iostream>
> #include <limits>
> ```

BTW: Please confirm that this is required.

---

*[drake/examples/Quadrotor/runDynamics.cpp, line 4 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KV_SYgd2nPXmylCYXmb:-KV_SYgd2nPXmylCYXmc:bak5mwd) ([raw file](https://github.com/robotlocomotion/drake/blob/483c97975c17db73809d5cd07fd27d78421cb489/drake/examples/Quadrotor/runDynamics.cpp#L4)):*
> ```cpp
> #include <limits>
> 
> #include <gflags/gflags.h>
> ```

A blank line is needed after this line.

---

*[drake/examples/Quadrotor/runDynamics.cpp, line 83 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KV_TgvZZL_E0pmYAM1n:-KV_TgvZZL_E0pmYAM1o:b-s4snpy) ([raw file](https://github.com/robotlocomotion/drake/blob/483c97975c17db73809d5cd07fd27d78421cb489/drake/examples/Quadrotor/runDynamics.cpp#L83)):*
> ```cpp
>         this->GetMutableSubsystemContext(context, plant_);
>     Eigen::VectorXd x0 = Eigen::VectorXd::Zero(12, 1);
>     x0(2) = 0.2;
> ```

Please provide a comment explaining what this state implies. For example, does this mean the quadrotor is initially 0.2m above the ground, facing forward, parallel with the ground, and not moving?

---

*[drake/examples/Quadrotor/runDynamics.cpp, line 90 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KV_T9QtPcyco1w9_-17:-KV_T9QtPcyco1w9_-18:b-juwrdv) ([raw file](https://github.com/robotlocomotion/drake/blob/483c97975c17db73809d5cd07fd27d78421cb489/drake/examples/Quadrotor/runDynamics.cpp#L90)):*
> ```cpp
> 
>  private:
>   systems::RigidBodyPlant<T>* plant_;
> ```

BTW, I think many of these class member variables could be local variables within the constructor of `Quadrotor::Quadrotor()`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-:-KV_U9rdEaXwrP7p6gw9:b44lg4b)*
<!-- Sent from Reviewable.io -->

foehnph commented:
@liangfok thanks, fixed it. PTAL.

---

Review status: 0 of 2 files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/examples/Quadrotor/CMakeLists.txt, line 5 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KV_SIGsURj-Ick5Cjtj:-KV_U_7sB2cXZIMmS5H4:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/483c97975c17db73809d5cd07fd27d78421cb489/drake/examples/Quadrotor/CMakeLists.txt#L5)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW: nit: Extra space before closing ")".
</details>
Done.

---

*[drake/examples/Quadrotor/CMakeLists.txt, line 6 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KV_SMFMiAZcd_YeNZKA:-KV_U_kP15H8hMdq56aI:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/483c97975c17db73809d5cd07fd27d78421cb489/drake/examples/Quadrotor/CMakeLists.txt#L6)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Typo: "mediu"
</details>
Done.

---

*[drake/examples/Quadrotor/runDynamics.cpp, line 1 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KV_SfXedIGWPOVewdHB:-KV_WCWyDD8XeciLeOMu:b-bokble) ([raw file](https://github.com/robotlocomotion/drake/blob/483c97975c17db73809d5cd07fd27d78421cb489/drake/examples/Quadrotor/runDynamics.cpp#L1)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW: Please confirm that this is required.
</details>
Done. deleted.

---

*[drake/examples/Quadrotor/runDynamics.cpp, line 2 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KV_SlFP1LiqOYdyO7MF:-KV_WEot_FM5FxMOWUsN:b-bokble) ([raw file](https://github.com/robotlocomotion/drake/blob/483c97975c17db73809d5cd07fd27d78421cb489/drake/examples/Quadrotor/runDynamics.cpp#L2)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW: Please confirm that this is required.
</details>
Done. deleted.

---

*[drake/examples/Quadrotor/runDynamics.cpp, line 4 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KV_SYgd2nPXmylCYXmb:-KV_UhOhkP4-hXIjDnVr:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/483c97975c17db73809d5cd07fd27d78421cb489/drake/examples/Quadrotor/runDynamics.cpp#L4)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> A blank line is needed after this line.
</details>
Done.

---

*[drake/examples/Quadrotor/runDynamics.cpp, line 83 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KV_TgvZZL_E0pmYAM1n:-KV_VUIJLEd1g-bxRtN-:b-cyb7gi) ([raw file](https://github.com/robotlocomotion/drake/blob/483c97975c17db73809d5cd07fd27d78421cb489/drake/examples/Quadrotor/runDynamics.cpp#L83)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Please provide a comment explaining what this state implies. For example, does this mean the quadrotor is initially 0.2m above the ground, facing forward, parallel with the ground, and not moving?
</details>
Done. yes it means exactly that.

---

*[drake/examples/Quadrotor/runDynamics.cpp, line 90 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KV_T9QtPcyco1w9_-17:-KV_aEY5gHaL5ZOmbGG1:bf9c5x3) ([raw file](https://github.com/robotlocomotion/drake/blob/483c97975c17db73809d5cd07fd27d78421cb489/drake/examples/Quadrotor/runDynamics.cpp#L90)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW, I think many of these class member variables could be local variables within the constructor of `Quadrotor::Quadrotor()`.
</details>
Done. only kept plant_ and lcm_

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-:-KV_aTLIOEAn7lw8AZ2F:b57fbg9)*
<!-- Sent from Reviewable.io -->

liangfok commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-:-KV_idxQTycoBb73vFPu:bnfp4nl)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
CI failure looks legit. PTAL.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[drake/examples/Quadrotor/runDynamics.cpp, line 45 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KVdTxsDYHvQAMbaXWvM:-KVdTxsDYHvQAMbaXWvN:b-i9wvuz) ([raw file](https://github.com/robotlocomotion/drake/blob/52ec1fb461c0779fdc437f75baaf0f9b854f21f1/drake/examples/Quadrotor/runDynamics.cpp#L45)):*
> ```cpp
>         builder.template AddSystem<systems::RigidBodyPlant<T>>(std::move(tree));
> 
>     Eigen::VectorXd hover_input = Eigen::VectorXd::Zero(4);
> ```

This might have to be VectorX<T> instead of Eigen::VectorXd

---

*[drake/examples/Quadrotor/runDynamics.cpp, line 82 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KVdUQeAAiNYoU0Do2AN:-KVdUQeAAiNYoU0Do2AO:b-b2gp9c) ([raw file](https://github.com/robotlocomotion/drake/blob/52ec1fb461c0779fdc437f75baaf0f9b854f21f1/drake/examples/Quadrotor/runDynamics.cpp#L82)):*
> ```cpp
>     systems::Context<T>* plant_context =
>         this->GetMutableSubsystemContext(context, plant_);
>     Eigen::VectorXd x0 = Eigen::VectorXd::Zero(12, 1);
> ```

As above...This might have to be VectorX<T> instead of Eigen::VectorXd 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-:-KVdUVNt3lKA3aHEehcs:b-4la9sl)*
<!-- Sent from Reviewable.io -->

foehnph commented:
thanks, already done! PTAL.

---

Review status: 1 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/Quadrotor/runDynamics.cpp, line 45 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KVdTxsDYHvQAMbaXWvM:-KVdVBAPKMuQL_R7phl0:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/52ec1fb461c0779fdc437f75baaf0f9b854f21f1/drake/examples/Quadrotor/runDynamics.cpp#L45)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary>
> This might have to be VectorX<T> instead of Eigen::VectorXd
</details>
Done.

---

*[drake/examples/Quadrotor/runDynamics.cpp, line 82 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KVdUQeAAiNYoU0Do2AN:-KVdVBz5mSPchC1eGyoY:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/52ec1fb461c0779fdc437f75baaf0f9b854f21f1/drake/examples/Quadrotor/runDynamics.cpp#L82)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary>
> As above...This might have to be VectorX<T> instead of Eigen::VectorXd 
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-:-KVdVRlhW7r7iuV7ok6y:b54ifys)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> pending CI result.

---

Review status: 1 of 2 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-:-KVdb6C3bX5nZT-345Bb:b-qv999t)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
+@jwnimmer-tri for platform review.

---

Review status: 1 of 2 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-:-KVdbG6Rpo0jBOpMyRfK:b-trzqz1)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
@foehnph Thanks for the help on this!

I will take a look later today, but for now one comment that would be great to fix first.

---

Reviewed 1 of 2 files at r5.
Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/Quadrotor/runDynamics.cpp, line 115 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KVducFVvLLG-SfBNI-W:-KVducFVvLLG-SfBNI-X:b-9qv67c) ([raw file](https://github.com/robotlocomotion/drake/blob/7f547858168b15107affd1f939cf476f7dd5d266/drake/examples/Quadrotor/runDynamics.cpp#L115)):*
> ```cpp
> int main(int argc, char* argv[]) {
>   return drake::examples::quadrotor::do_main(argc, argv);
> }
> ```

_"Filenames should be all lowercase and can include underscores"_
_"C++ files should end in .cc"_
https://google.github.io/styleguide/cppguide.html#File_Names
Lots of old code is not yet updated per that rule, but new PRs must follow it.

Also, the filename and binary program name should be identical (modulo the `.cc` extension).

Those would suggest something like `run_quadrotor_dynamics.cc` for the filename, but of course the specific choice is up to you.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-:-KVduZDUtSMRO_d4zYAF:b-26sp0f)*
<!-- Sent from Reviewable.io -->

foehnph commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/Quadrotor/run_quadrotor_dynamics.cc, line 115 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KVducFVvLLG-SfBNI-W:-KVeBPMoTwJFXlmyq_nC:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/7f547858168b15107affd1f939cf476f7dd5d266/drake/examples/Quadrotor/run_quadrotor_dynamics.cc#L115)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> _"Filenames should be all lowercase and can include underscores"_
> _"C++ files should end in .cc"_
> https://google.github.io/styleguide/cppguide.html#File_Names
> Lots of old code is not yet updated per that rule, but new PRs must follow it.
> 
> Also, the filename and binary program name should be identical (modulo the `.cc` extension).
> 
> Those would suggest something like `run_quadrotor_dynamics.cc` for the filename, but of course the specific choice is up to you.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3995)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 3 of 3 files at r7.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/examples/Quadrotor/CMakeLists.txt, line 4 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KVeDpmRX1rpGbCS_KCY:-KVeDpmRX1rpGbCS_KCZ:b-l2wiqy) ([raw file](https://github.com/robotlocomotion/drake/blob/f043dc673f04187006aae14a4cfdd39e6a611636/drake/examples/Quadrotor/CMakeLists.txt#L4)):*
> ```CMake
> # due to a bug in Eigen; see issue #2106 and PR #2107.
> if(lcm_FOUND AND NOT (WIN32 AND (CMAKE_SIZEOF_VOID_P EQUAL 4)))
>   add_executable(runQuadrotorDynamics run_quadrotor_dynamics.cc)
> ```

The instances of `runQuadrotorDynamics` in this file should now be `run_quadrotor_dynamics`, so that the program name also changes to match the source filename.

---

*[drake/examples/Quadrotor/run_quadrotor_dynamics.cc, line 55 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KVeGY8f3Ygyn5E6iWH9:-KVeGY8ge52Hw4G89fJC:b-ot2r6j) ([raw file](https://github.com/robotlocomotion/drake/blob/f043dc673f04187006aae14a4cfdd39e6a611636/drake/examples/Quadrotor/run_quadrotor_dynamics.cc#L55)):*
> ```cpp
>             plant_->get_rigid_body_tree(), &lcm_);
> 
>     std::cout << plant_->get_input_size();
> ```

This output seems like it would perplex a user of this program.

---

*[drake/examples/Quadrotor/run_quadrotor_dynamics.cc, line 57 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KVeGuRiBMmSo1t41yMt:-KVeGuRiBMmSo1t41yMu:bkpvg45) ([raw file](https://github.com/robotlocomotion/drake/blob/f043dc673f04187006aae14a4cfdd39e6a611636/drake/examples/Quadrotor/run_quadrotor_dynamics.cc#L57)):*
> ```cpp
>     std::cout << plant_->get_input_size();
>     builder.Connect(source->get_output_port(), plant_->get_input_port(0));
>     builder.Connect(plant_->get_output_port(0), publisher->get_input_port(0));
> ```

BTW For the above two lines, consider `builder.Cascade(source, *plant_);` and `builder.Cascade(*plant_, *publisher);`, which I believe is clearer and more robust in that only a single input and output port are relevant for each system (and it will throw if there is more than one port).  In this code, the `0`s are a bit of "magic" numbers and could bitrot.

---

*[drake/examples/Quadrotor/run_quadrotor_dynamics.cc, line 62 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KVeF5OXMWpMhVqBBkN8:-KVeF5OY2qWcoSwR8PM_:b-gb1u74) ([raw file](https://github.com/robotlocomotion/drake/blob/f043dc673f04187006aae14a4cfdd39e6a611636/drake/examples/Quadrotor/run_quadrotor_dynamics.cc#L62)):*
> ```cpp
>   }
> 
>   void AddGround(RigidBodyTree* tree) {
> ```

BTW Consider adding a `TODO` atop this method reminding us that this function is duplicated widely throughout the codebase, and should be refactored into a single common location.

---

*[drake/examples/Quadrotor/run_quadrotor_dynamics.cc, line 86 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KVeKsUJ4N0KaPb81I-I:-KVeKsUJ4N0KaPb81I-J:bwc4mcl) ([raw file](https://github.com/robotlocomotion/drake/blob/f043dc673f04187006aae14a4cfdd39e6a611636/drake/examples/Quadrotor/run_quadrotor_dynamics.cc#L86)):*
> ```cpp
>     VectorX<T> x0(plant_->get_num_states());
>     x0.setZero();
>     x0(2) = 0.2;  // Resting at 0.2 m above origin, horizontally forward.
> ```

Unclear where magic "2" comes from.

Is it related to the links in the urdf?  Would this program fail-fast if the urdf was editing without fixing this magic number?  Is there a way to check that "2" means what we think it means, relative to the `RigidBodyTree`'s by-name accessors?

---

*[drake/examples/Quadrotor/run_quadrotor_dynamics.cc, line 91 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KVeJmd6bk1lPAR7OJOp:-KVeJmd6bk1lPAR7OJOq:b-axovte) ([raw file](https://github.com/robotlocomotion/drake/blob/f043dc673f04187006aae14a4cfdd39e6a611636/drake/examples/Quadrotor/run_quadrotor_dynamics.cc#L91)):*
> ```cpp
> 
>  private:
>   systems::RigidBodyPlant<T>* plant_;
> ```

Missing `{}` on POD member data, i.e., `... plant_{};`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3995)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 1 of 3 files at r7.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3995)*
<!-- Sent from Reviewable.io -->

foehnph commented:
fixed, PTAL, especially at `Connect()`/`Cascade()` comment.
I will go on and rebase, then commit again.

---

Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/examples/Quadrotor/CMakeLists.txt, line 4 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KVeDpmRX1rpGbCS_KCY:-KVe_kQk3kIwqQb6Dsig:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f043dc673f04187006aae14a4cfdd39e6a611636/drake/examples/Quadrotor/CMakeLists.txt#L4)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> The instances of `runQuadrotorDynamics` in this file should now be `run_quadrotor_dynamics`, so that the program name also changes to match the source filename.
</details>
Done.

---

*[drake/examples/Quadrotor/run_quadrotor_dynamics.cc, line 55 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KVeGY8f3Ygyn5E6iWH9:-KVe_6MkdBFkhrs2pqwk:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f043dc673f04187006aae14a4cfdd39e6a611636/drake/examples/Quadrotor/run_quadrotor_dynamics.cc#L55)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> This output seems like it would perplex a user of this program.
</details>
Done.

---

*[drake/examples/Quadrotor/run_quadrotor_dynamics.cc, line 57 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KVeGuRiBMmSo1t41yMt:-KVikGzY-a846cg5MVPi:b-l5ddb) ([raw file](https://github.com/robotlocomotion/drake/blob/f043dc673f04187006aae14a4cfdd39e6a611636/drake/examples/Quadrotor/run_quadrotor_dynamics.cc#L57)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW For the above two lines, consider `builder.Cascade(source, *plant_);` and `builder.Cascade(*plant_, *publisher);`, which I believe is clearer and more robust in that only a single input and output port are relevant for each system (and it will throw if there is more than one port).  In this code, the `0`s are a bit of "magic" numbers and could bitrot.
</details>
Not done since `builder.Cascade(*plant_, *publisher);` is not possible.
`plant_`has two output ports (state vector and KinematicsResult), whereas `Cascade()`expects one output port.
Shall I do both with `Connect()` or only the first one with `Cascade()`?

---

*[drake/examples/Quadrotor/run_quadrotor_dynamics.cc, line 62 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KVeF5OXMWpMhVqBBkN8:-KVe_LXDqW8CYXXER3f2:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f043dc673f04187006aae14a4cfdd39e6a611636/drake/examples/Quadrotor/run_quadrotor_dynamics.cc#L62)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW Consider adding a `TODO` atop this method reminding us that this function is duplicated widely throughout the codebase, and should be refactored into a single common location.
</details>
Done.

---

*[drake/examples/Quadrotor/run_quadrotor_dynamics.cc, line 86 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KVeKsUJ4N0KaPb81I-I:-KVik7PQimgy4GBvHuU3:b-eyfq57) ([raw file](https://github.com/robotlocomotion/drake/blob/f043dc673f04187006aae14a4cfdd39e6a611636/drake/examples/Quadrotor/run_quadrotor_dynamics.cc#L86)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Unclear where magic "2" comes from.
> 
> Is it related to the links in the urdf?  Would this program fail-fast if the urdf was editing without fixing this magic number?  Is there a way to check that "2" means what we think it means, relative to the `RigidBodyTree`'s by-name accessors?
</details>
Done. added more precise description

---

*[drake/examples/Quadrotor/run_quadrotor_dynamics.cc, line 91 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KVeJmd6bk1lPAR7OJOp:-KVe_UtXeOMk-VbufL86:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f043dc673f04187006aae14a4cfdd39e6a611636/drake/examples/Quadrotor/run_quadrotor_dynamics.cc#L91)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Missing `{}` on POD member data, i.e., `... plant_{};`.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-:-KVisHJPjfpK4d-HKH1f:b-qn57ur)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 3 files at r7, 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/Quadrotor/run_quadrotor_dynamics.cc, line 57 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KVeGuRiBMmSo1t41yMt:-KVj1qN1m5UTw4JQ921Z:bbx5lcn) ([raw file](https://github.com/robotlocomotion/drake/blob/f043dc673f04187006aae14a4cfdd39e6a611636/drake/examples/Quadrotor/run_quadrotor_dynamics.cc#L57)):*
<details><summary><i>Previously, foehnph (Philipp Föhn) wrote…</i></summary>
> Not done since `builder.Cascade(*plant_, *publisher);` is not possible.
> `plant_`has two output ports (state vector and KinematicsResult), whereas `Cascade()`expects one output port.
> Shall I do both with `Connect()` or only the first one with `Cascade()`?
</details>
Alright.  It's a defect of the plant that doesn't offer a named getter for its output, but that's not a problem to fix in this PR.  Marking resolved.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-:-KVj22fMXhkHZqs1sb_X:bnfp4nl)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 1 of 3 files at r7, 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/Quadrotor/run_quadrotor_dynamics.cc, line 61 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KVjudVqLpdLWOaG69wz:-KVjudVqLpdLWOaG69x-:b-n790nr) ([raw file](https://github.com/robotlocomotion/drake/blob/bc2712f1ec8aa4a1df7347c8a7bfa2ec83b7f5fd/drake/examples/Quadrotor/run_quadrotor_dynamics.cc#L61)):*
> ```cpp
>   }
> 
>   // TODO(foehnph): dublicated, refactor into single location
> ```

Typo: "dublicated". Missing period at end of sentence.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3995)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 3 files at r7, 1 of 1 files at r10.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3995)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 1 of 3 files at r7, 1 of 1 files at r10.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3995)*
<!-- Sent from Reviewable.io -->

foehnph commented:
credits to @naveenoid for the rebase!
changed the typo and a wrong command in the makefile

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/Quadrotor/run_quadrotor_dynamics.cc, line 61 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-KVjudVqLpdLWOaG69wz:-KVoB21zL6br46ly1zfF:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/bc2712f1ec8aa4a1df7347c8a7bfa2ec83b7f5fd/drake/examples/Quadrotor/run_quadrotor_dynamics.cc#L61)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Typo: "dublicated". Missing period at end of sentence.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3995#-:-KVjzobb3Jfmsv8an-8b:b-ubdiir)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 3 files at r7, 2 of 2 files at r11.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3995)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 1 of 3 files at r7, 2 of 2 files at r11.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3995)*
<!-- Sent from Reviewable.io -->

foehnph commented:
@drake-jenkins-bot linux-gcc-experimental-ros please

liangfok commented:
The [previous linux-gcc-experimental-ros CI test](https://drake-jenkins.csail.mit.edu/job/linux-gcc-experimental-ros/532/console) was aborted due to extremely high latency (the underlying cause is a mystery to me):

```
[drake:make] /usr/bin/make -f src/signal_scope/CMakeFiles/signal-scoBuild timed out (after 240 minutes). Marking the build as aborted.
```

@drake-jenkins-bot linux-gcc-experimental-ros please

liangfok commented:
Looks like this PR is ready to be merged. Congrats!
